### DISCUSSION TITLE
Frozen binary on Linux doesn't work if source python is not present on the system

### DISCUSSION THREAD

**Discussion by HakanSunay** [marcelotduarte/cx_Freeze#1886] [Q&A] [answered]
Hello,

We have a moderately complex application that is packaged for Windows, Mac, and Linux.
For Linux, there seems to be a problem, when we execute the frozen binary on a system that does not have `python3.10` (the application was frozen using `python3.10`).
This is the error we get, when we run on Ubuntu-18.04.2 where `python3.10` is missing:
```
$ <my_binary> --help
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = (not set)
  program name = 'python3'
  isolated = 0
  environment = 0
  user site = 1
  import site = 0
  sys._base_executable = '/usr/bin/python3'
  sys.base_prefix = '<redacted>'
  sys.base_exec_prefix = '<redacted>'
  sys.platlibdir = 'lib'
  sys.executable = '/usr/bin/python3'
  sys.prefix = '<redacted>'
  sys.exec_prefix = '<redacted>'
  sys.path = [
    '<redacted>/lib/python310.zip',
    '<redacted>/lib/python3.10',
    '<redacted>lib/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007f0fa3
```

The `<redacted>` part is a location in our corporate build farm, where Python was initially built from source.
This location is not accessible by any machine.

`strace` details seem to indicate that the Python of the host machine is inspected explicitly for `python3.10`:
```
stat("/usr/local/sbin/python3", 0x7ffd82d76a00) = -1 ENOENT (No such file or directory)
stat("/usr/local/bin/python3", 0x7ffd82d76a00) = -1 ENOENT (No such file or directory)
stat("/usr/sbin/python3", 0x7ffd82d76a00) = -1 ENOENT (No such file or directory)
stat("/usr/bin/python3", {st_mode=S_IFREG|0755, st_size=4526456, ...}) = 0
readlink("/usr/bin/python3", "python3.6", 4096) = 9
readlink("/usr/bin/python3.6", 0x7ffd82d718e0, 4096) = -1 EINVAL (Invalid argument)
openat(AT_FDCWD, "/usr/bin/pyvenv.cfg", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/pyvenv.cfg", O_RDONLY) = -1 ENOENT (No such file or directory)
stat("/usr/bin/Modules/Setup.local", 0x7ffd82d72880) = -1 ENOENT (No such file or directory)
stat("/usr/bin/lib/python3.10/os.py", 0x7ffd82d72790) = -1 ENOENT (No such file or directory)
stat("/usr/bin/lib/python3.10/os.pyc", 0x7ffd82d72790) = -1 ENOENT (No such file or directory)
stat("/usr/lib/python3.10/os.py", 0x7ffd82d72790) = -1 ENOENT (No such file or directory)
stat("/usr/lib/python3.10/os.pyc", 0x7ffd82d72790) = -1 ENOENT (No such file or directory)
stat("<redacted>/lib/python3.10/os.py", 0x7ffd82d72790) = -1 ENOENT (No such file or directory)
stat("<redacted>/lib/python3.10/os.pyc", 0x7ffd82d72790) = -1 ENOENT (No such file or directory)
```
and once that is not found, the first message is printed out.

I am trying to understand if this is a known issue, and if there is a way to overcome this by somehow packaging the right Python so that the binary can find it under `lib64/lin` or somewhere else.

We are consuming:
https://github.com/marcelotduarte/cx_Freeze/commit/02081348b4d518fcba2aa91219eb00f03126a29d

**Comment by marcelotduarte** [comment, marked as answer]
Are you using cx_Freeze from sources?
Try to use from wheels instead:
`pip install -U cx_Freeze`

**Reply by HakanSunay** [reply]
Yes, we are using cx_Freeze from source. Actually, we are building it from source.
Do you think we can work around this problem without using `pip`?
.. and possibly with the same cx_Freeze version.

I am trying to understand what is not working well for Linux, what is missing and else.
Do you suspect an incomplete build process for cx_Freeze from source?

**Reply by marcelotduarte** [reply]
Probably the libpython (or another library) is missing in the distribution.
How you installed python in the build machine?
How you compiled cx_Freeze?
Why not follow the recommended instructions to install?

In ubuntu python should be installed using `build-essential` and `python3-dev` packages.

**Reply by HakanSunay** [reply]
> Probably the libpython (or another library) is missing in the distribution.

libpython is certainly there and many other, but there might be something missing indeed.
Is there a way to find what could be missing?

> How you installed python in the build machine?

python and cx_Freeze are built on separate build machines.
Then they are consumed by another build machine that builds my binary.
In the end, I am trying to run the binary on an arbitrary Linux distribution - in my case, that is Ubuntu 18.

>  How you compiled cx_Freeze?

In summary:
python3 setup.py build_ext
python3 setup.py build_py
python3 setup.py build_scripts
python3 setup.py build_clib
python3 setup.py install


>  Why not follow the recommended instructions to install?

We have certain restrictions for installing packages.

**Comment by HakanSunay** [comment]
We followed your recommendation and switched to using the wheel published in pypi, thanks.
